segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -209,7 +212,9 @@ public void commit(MetadataTransaction txn, boolean lazyWrite, boolean skipStore
         }
         // Step 2 : Check whether transaction is safe to commit.
         // This check needs to be atomic, with absolutely no possibility of re-entry
-        synchronized (lock) {
+        val writeLock = lockManager.writeLock(txn.getKeysToLock());
+        try {
+            writeLock.lock();
```

Comment:
This is performing IO while holding a lock which is generally considered bad form, and in our contributing guidelines it says:

> If a blocking operation is performed under lock, it should be very obvious and clear that it is absolutely necessary.

From my point of view this criteria has not been met.

Additionally, can't there be concurrent modification from other processes?

Reply:
For example: Did you consider making this a 'service' which enqueues callso onto a queue and returns completable futures while a background pool performs the IO?

Reply:
Serializing transactions is necessary for consistency and also persisting these transactions in actual modifications in TableStore or as records in SystemJournal is also essential, transaction can not succeed if external operation fails. 

I do plan to convert methods in this class into async API in 0.9. Some of this code will change then.
But the main purpose of  this PR is to remove global lock and instead synchronize writes based on per segment lock.

Reply:
https://github.com/pravega/pravega/issues/5067